File: src/pl/swmud/ns/swaedit/gui/ToolTipEventFilter.java
package pl.swmud.ns.swaedit.gui;

import com.trolltech.qt.core.QEvent;
import com.trolltech.qt.core.QObject;
import com.trolltech.qt.gui.QHelpEvent;
import com.trolltech.qt.gui.QWidget;

class ToolTipEventFilter extends QObject {

    public boolean eventFilter(QObject o, QEvent e) {
        if (e.type() == QEvent.Type.ToolTip && o instanceof QWidget) {
            QHelpEvent he = (QHelpEvent)e;
            QWidget w = (QWidget)o;
            if (!w.toolTip().isEmpty() && BalloonWidget.canShowToolTip(w, he.globalPos())) {
                new BalloonWidget(he.globalPos()).showToolTip(w);
            }
            e.accept();
            return true;
        }
        else {
            return super.eventFilter(o, e);
        }

    }

}

File: src/gui/ToolTipEventFilter.cpp
#include "gui/ToolTipEventFilter.h"
#include "gui/BalloonWidget.h"

#include <QEvent>
#include <QHelpEvent>
#include <QWidget>

ToolTipEventFilter::ToolTipEventFilter(QObject *parent)
    : QObject(parent)
{
}

bool ToolTipEventFilter::eventFilter(QObject *watched, QEvent *event)
{
    if (event->type() == QEvent::ToolTip) {
        QWidget *w = qobject_cast<QWidget *>(watched);
        if (w) {
            QHelpEvent *he = static_cast<QHelpEvent *>(event);
            if (!w->toolTip().isEmpty() && BalloonWidget::canShowToolTip(w, he->globalPos())) {
                BalloonWidget *balloon = new BalloonWidget(he->globalPos());
                balloon->showToolTip(w);
            }
            event->accept();
            return true;
        }
    }

    return QObject::eventFilter(watched, event);
}
